Poll for closed state, rather than waiting fixed time

illicitonion opened the pull request.

illicitonion committed:
Add fork me ribon, mention gitrust in readme

illicitonion committed:
Poll for closed state, rather than waiting fixed time

illicitonion committed:
Compensate for possibly-empty commit users

If a commit was made by an author whose email address isn't linked to a
github account, there is no author object in the commit.

There is, however, a commit object in the commit, which contains a
(different) author object with non-account-related metadata. Grab some
data from there instead.

This was found when NegativeMjark tried to use review.rocks, and it
turns out one of his commits from one of his PRs was not associated with
an account. He conveniently provided me with a har file which made
tracking this down pretty easy. I've added that example PR (its first
commit is lacking an author) to examples, as well as a script for
extracting github responses from har files, because I'm sure it will be
useful again.

illicitonion committed:
Use a slightly more sane URL

illicitonion committed:
Add delay between polls

We actually wait a really long time

kegsay committed:
Merge pull request #64 from illicitonion/forkme

Merge pull request #64 from illicitonion/forkme

kegsay committed:
Merge pull request #66 from illicitonion/unknownusers

Merge pull request #66 from illicitonion/unknownusers

kegsay committed:
Merge pull request #67 from illicitonion/loginprompt

Note about needing to log in to merge

kegsay commented on lib/logic/controller.js:
```diff
@@ -228,6 +228,26 @@ Controller.prototype.open = function(pr) {
+        () => !closed,
+        () => this.httpApi.getPullRequest(pr.getRepo(), pr.getId())
+                .then((res) => { closed = res.body.state === "closed"; }),
+        200
```
Way too aggressive timings. The HTTP requests will be quick so this will spam a lot. Minimum 1 second, preferably 3. Ditto `:237`

illicitonion replied:
Done - bumped both; closed seems to actually happen pretty quickly, but mergeable is pretty slow

kegsay commented on lib/logic/controller.js:
```diff
@@ -242,4 +262,26 @@ Controller.prototype._handleMergeError = function(err) {
+ * _while returns a promise which is resolved when condition() is true, repeatedly calling fn()
+ * between checks of condition().
+ */
+Controller.prototype._while = function(condition, fn, opt_delay_ms) {
```
This is cute. Perhaps this should be put into a util class? We have one in `/components/utils.js` which should probably be moved up to `/utils.js`.

illicitonion replied:
Done

kegsay committed:
Allow anchors on non BASE->HEAD diffs

This is horrible code. Once we have tests for anchors, this needs
to be sanitised.

illicitonion committed:
Merge branch 'poll' of github.com:illicitonion/github-pull-review into poll

illicitonion committed:
Bump timings up

illicitonion committed:
Pull out while into utils

illicitonion committed:
Bump timeout more

Kegsay commented:
LGTM
